File: scraper_service/hack_scraper/date_utils.py
import re
from datetime import datetime, date

MONTHS = ['january', 'february', 'march', 'april', 'may', 'june', 'july', 'august', 'september', 'october', 'november', 'december']
MONTH_SHORT = ['jan', 'feb', 'mar', 'apr', 'may', 'jun', 'jul', 'aug', 'sep', 'oct', 'nov', 'dec']

def parse_deadline_date(deadline_str):
    """
    Parses various deadline string formats into a datetime.date object.
    
    Supports:
    - "2026-12-31" (ISO)
    - "Jul 15, 2026" / "15 Aug 2026"
    - "Mar 10 - 11, 2026" (Range with implicit month on right side)
    - "Jun 01 - Jul 15, 2026" (Extracts end date)
    - "Aug 10, 2026 @ 5:00pm EDT"
    - "21st Jul 2026" / "Starts 20 Jul 2026"
    """
    if not deadline_str or not isinstance(deadline_str, str):
        return None

    s = deadline_str.strip()
    if not s:
        return None

    lower = s.lower()
    if any(k in lower for k in ["ended", "closed", "finished"]):
        return None

    # Handle range (e.g. "Mar 10 - 11, 2026" or "Jun 01 - Jul 15, 2026")
    if " - " in s or ("-" in s and not re.match(r"^\d{4}-\d{2}-\d{2}$", s)):
        parts = re.split(r" - |\bto\b", s, flags=re.IGNORECASE)
        left_part = parts[0].strip()
        right_part = parts[-1].strip()

        right_lower = right_part.lower()
        has_month = any(m in right_lower for m in MONTH_SHORT)

        if not has_month:
            match = re.search(r"([a-zA-Z]{3,9})", left_part)
            if match:
                right_part = f"{match.group(1)} {right_part}"
        s = right_part

    # Clean up noise
    s = re.sub(r"@.*$", "", s).strip()
    s = re.sub(r"^starts\s+", "", s, flags=re.IGNORECASE).strip()
    s = re.sub(r"(\d+)(st|nd|rd|th)", r"\1", s, flags=re.IGNORECASE).strip()

    # Try dateutil if installed
    try:
        from dateutil import parser
        parsed_dt = parser.parse(s, fuzzy=True)
        return parsed_dt.date()
    except Exception:
        pass

    # Standard strptime fallback formats
    formats = [
        "%Y-%m-%d",
        "%b %d, %Y",
        "%B %d, %Y",
        "%d %b %Y",
        "%d %B %Y",
        "%b %d %Y",
        "%B %d %Y",
        "%Y/%m/%d",
        "%m/%d/%Y",
    ]

    for fmt in formats:
        try:
            return datetime.strptime(s, fmt).date()
        except ValueError:
            continue

    # Try appending current year if year missing (e.g., "Jul 15")
    current_year = date.today().year
    s_with_year = f"{s}, {current_year}"
    for fmt in ["%b %d, %Y", "%B %d, %Y", "%d %b, %Y", "%d %B, %Y"]:
        try:
            return datetime.strptime(s_with_year, fmt).date()
        except ValueError:
            continue

    return None

def is_title_or_text_from_past(text, reference_date=None):
    """
    Checks if text (e.g., title) contains a year or month/year before reference_date.
    """
    if not text or not isinstance(text, str):
        return False

    if reference_date is None:
        reference_date = date.today()

    current_year = reference_date.year

    # Check 4-digit years or 2K years e.g. 2012, 2013, 2K12
    year_match = re.search(r"\b(20[0-9]{2}|2K[0-9]{2})\b", text, flags=re.IGNORECASE)
    if year_match:
        y_str = year_match.group(1).replace("2K", "20").replace("2k", "20")
        try:
            y = int(y_str)
            if y < current_year:
                return True
        except ValueError:
            pass

    # Check month + year e.g. "March 2026", "Feb 2026"
    month_year_match = re.search(
        r"\b(jan(?:uary)?|feb(?:ruary)?|mar(?:ch)?|apr(?:il)?|may|jun(?:e)?|jul(?:y)?|aug(?:ust)?|sep(?:tember)?|oct(?:ober)?|nov(?:ember)?|dec(?:ember)?)\s*(\d{1,2})?,?\s*(20[0-9]{2})\b",
        text,
        flags=re.IGNORECASE
    )
    if month_year_match:
        month_str = month_year_match.group(1).lower()
        day = int(month_year_match.group(2)) if month_year_match.group(2) else 28
        year = int(month_year_match.group(3))

        month_idx = -1
        for idx, m in enumerate(MONTHS):
            if m.startswith(month_str):
                month_idx = idx + 1
                break
        if month_idx == -1:
            for idx, m in enumerate(MONTH_SHORT):
                if m == month_str[:3]:
                    month_idx = idx + 1
                    break

        if month_idx != -1:
            try:
                dt = date(year, month_idx, day)
                if dt < reference_date:
                    return True
            except ValueError:
                pass

    return False

def is_hackathon_closed(deadline_str, title_str="", reference_date=None):
    """
    Checks if a hackathon's deadline or title indicates it is closed/expired compared to reference_date.
    """
    if reference_date is None:
        reference_date = date.today()

    if is_title_or_text_from_past(title_str, reference_date):
        return True

    if not deadline_str or not isinstance(deadline_str, str):
        return False

    lower = deadline_str.strip().lower()
    if any(k in lower for k in ["ended", "closed", "finished"]):
        return True

    if is_title_or_text_from_past(deadline_str, reference_date):
        return True

    deadline_dt = parse_deadline_date(deadline_str)
    if not deadline_dt:
        return False

    return deadline_dt < reference_date

File: backend/src/utils/dateUtils.js
/**
 * Utility functions for parsing hackathon deadlines and checking if registration/hackathon is closed or finished.
 */

const MONTHS = ['january', 'february', 'march', 'april', 'may', 'june', 'july', 'august', 'september', 'october', 'november', 'december'];
const MONTH_SHORT = ['jan', 'feb', 'mar', 'apr', 'may', 'jun', 'jul', 'aug', 'sep', 'oct', 'nov', 'dec'];

/**
 * Parses various deadline string formats into a JavaScript Date object.
 * Handles formats like:
 * - "2026-12-31" (ISO)
 * - "Jul 15, 2026" / "15 Aug 2026"
 * - "Mar 10 - 11, 2026" (Range with implicit month on right side)
 * - "Jun 01 - Jul 15, 2026" (Range with explicit month on right side)
 * - "Aug 10, 2026 @ 5:00pm EDT"
 * - "21st Jul 2026" / "Starts 20 Jul 2026"
 * 
 * @param {string} deadlineStr - Raw deadline string
 * @returns {Date|null} - Valid Date object or null
 */
function parseDeadlineDate(deadlineStr) {
  if (!deadlineStr || typeof deadlineStr !== 'string') return null;

  let str = deadlineStr.trim();
  if (!str) return null;

  const lower = str.toLowerCase();
  if (lower.includes('ended') || lower.includes('closed') || lower.includes('finished')) {
    return null;
  }

  // Handle date ranges e.g. "Mar 10 - 11, 2026" or "Jun 01 - Jul 15, 2026"
  if (str.includes(' - ') || (str.includes('-') && !/^\d{4}-\d{2}-\d{2}$/.test(str))) {
    const parts = str.split(/-|\bto\b/i);
    const leftPart = parts[0].trim();
    let rightPart = parts[parts.length - 1].trim();

    // Check if rightPart has a month
    const rightLower = rightPart.toLowerCase();
    const hasMonth = MONTH_SHORT.some(m => rightLower.includes(m));

    if (!hasMonth) {
      // Find month from leftPart
      const match = leftPart.match(/([a-zA-Z]{3,9})/);
      if (match) {
        rightPart = match[1] + ' ' + rightPart;
      }
    }
    str = rightPart;
  }

  // Clean up noise
  str = str.replace(/@.*$/, '').trim();
  str = str.replace(/^starts\s+/i, '').trim();
  str = str.replace(/(\d+)(st|nd|rd|th)/gi, '$1');

  // Direct Date parse attempt
  let parsedDate = new Date(str);
  if (!isNaN(parsedDate.getTime())) {
    return parsedDate;
  }

  // Try appending current year if year missing (e.g. "Jul 15" -> "Jul 15, 2026")
  const currentYear = new Date().getFullYear();
  parsedDate = new Date(`${str}, ${currentYear}`);
  if (!isNaN(parsedDate.getTime())) {
    return parsedDate;
  }

  return null;
}

/**
 * Checks if a title or text explicitly specifies a past date/year (e.g. "Tangelo Town 2013", "March 2026" when reference is July 2026).
 * 
 * @param {string} text - Title or raw text
 * @param {Date} referenceDate - Current reference date
 * @returns {boolean} - true if text refers to a past date/year
 */
function isTitleOrTextFromPast(text, referenceDate) {
  if (!text || typeof text !== 'string') return false;

  const currentYear = referenceDate.getFullYear();

  // Match 4-digit years e.g. 2012, 2013, 2024, 2025
  const yearMatch = text.match(/\b(20[0-9]{2}|2K[0-9]{2})\b/i);
  if (yearMatch) {
    const year = parseInt(yearMatch[1].replace(/2K/i, '20'));
    if (year < currentYear) {
      return true;
    }
  }

  // Match month + year in text e.g. "March 2026", "Feb 2026"
  const monthYearMatch = text.match(/\b(jan(?:uary)?|feb(?:ruary)?|mar(?:ch)?|apr(?:il)?|may|jun(?:e)?|jul(?:y)?|aug(?:ust)?|sep(?:tember)?|oct(?:ober)?|nov(?:ember)?|dec(?:ember)?)\s*(\d{1,2})?,?\s*(20[0-9]{2})\b/i);
  if (monthYearMatch) {
    const monthStr = monthYearMatch[1].toLowerCase();
    const day = monthYearMatch[2] ? parseInt(monthYearMatch[2]) : 28;
    const year = parseInt(monthYearMatch[3]);

    let monthIdx = MONTHS.findIndex(m => m.startsWith(monthStr));
    if (monthIdx === -1) monthIdx = MONTH_SHORT.findIndex(m => m === monthStr.slice(0, 3));

    if (monthIdx !== -1) {
      const dt = new Date(year, monthIdx, day, 23, 59, 59);
      if (dt.getTime() < referenceDate.getTime()) {
        return true;
      }
    }
  }

  return false;
}

/**
 * Determines whether a hackathon is closed or finished based on its deadline string, title, and reference date.
 * 
 * @param {string} deadlineStr - The deadline string
 * @param {string} [titleStr] - Optional title of the hackathon
 * @param {Date} [referenceDate] - Optional date to compare against (defaults to now)
 * @returns {boolean} - true if the deadline or title indicates it is closed/expired
 */
function isHackathonClosed(deadlineStr, titleStr = '', referenceDate = new Date()) {
  const refStartOfDay = new Date(referenceDate);
  refStartOfDay.setHours(0, 0, 0, 0);

  // Check title for past years/months
  if (isTitleOrTextFromPast(titleStr, refStartOfDay)) {
    return true;
  }

  if (!deadlineStr || typeof deadlineStr !== 'string') return false;

  const lower = deadlineStr.trim().toLowerCase();
  if (lower.includes('ended') || lower.includes('closed') || lower.includes('finished')) {
    return true;
  }

  // Check deadlineStr for past years/months
  if (isTitleOrTextFromPast(deadlineStr, refStartOfDay)) {
    return true;
  }

  const deadlineDate = parseDeadlineDate(deadlineStr);
  if (!deadlineDate) return false;

  // Set deadline date to end of day if no time component was provided (23:59:59)
  const deadlineEndOfDay = new Date(deadlineDate);
  if (deadlineEndOfDay.getHours() === 0 && deadlineEndOfDay.getMinutes() === 0) {
    deadlineEndOfDay.setHours(23, 59, 59, 999);
  }

  return deadlineEndOfDay.getTime() < refStartOfDay.getTime();
}

module.exports = {
  parseDeadlineDate,
  isTitleOrTextFromPast,
  isHackathonClosed,
};
